core/src/main/scala/org/apache/spark/deploy/rest/StandaloneRestClient.scala

```
@@ -63,45 +71,109 @@ private[deploy] class StandaloneRestClient extends Logging {
+          if(handleConnectionException(m)) {
+            throw new SubmitRestConnectionException(
+              s"Unable to connect to server", unreachable)
+          }
+        case malformed @ (_: JsonProcessingException | _: SubmitRestProtocolException) =>
+          if(handleConnectionException(m)) {
+            throw new SubmitRestProtocolException(
+              "Malformed response received from server", malformed)
+          }
+      }
     }
     response
   }
 
   /** Request the status of a submission from the server. */
   def requestSubmissionStatus(
-      master: String,
       submissionId: String,
       quiet: Boolean = false): SubmitRestProtocolResponse = {
     logInfo(s"Submitting a request for the status of submission $submissionId in $master.")
-    validateMaster(master)
-    val response = get(getStatusUrl(master, submissionId))
-    response match {
-      case s: SubmissionStatusResponse => if (!quiet) { handleRestResponse(s) }
-      case unexpected => handleUnexpectedRestResponse(unexpected)
+    var suc: Boolean = false
+    var response: SubmitRestProtocolResponse = null
+    for (m <- masters) {
+      validateMaster(m)
+      val url = getStatusUrl(m, submissionId)
+      try {
+        response = get(url)
+        response match {
+          case s: SubmissionStatusResponse if s.success =>
+            if (!quiet) {
+              handleRestResponse(s)
+            }
+            suc = true
+          case unexpected =>
+            handleUnexpectedRestResponse(unexpected)
+        }
+      } catch {
+        case unreachable @ (_: FileNotFoundException | _: SocketException | _: ConnectException) =>
+          if(handleConnectionException(m)) {
+            throw new SubmitRestConnectionException(
+              s"Unable to connect to server", unreachable)
+          }
+        case malformed @ (_: JsonProcessingException | _: SubmitRestProtocolException) =>
+          if(handleConnectionException(m)) {
```

This is not a connection exception. We should just throw the exception here